include/odp/api/spec/classification.h

```diff
@@ -107,6 +108,55 @@ typedef union odp_cls_pmr_terms_t {
 	uint64_t all_bits;
 } odp_cls_pmr_terms_t;
 
+/** Random Early Detection (RED)
+ * Random Early Detection is enabled to initiate a drop probability
+ * for the incoming packet when the packets in the queue/pool reaches
+ * a specified threshold.
+ * When RED is enabled for a particular flow then further incoming
+ * packets are assigned a drop probability based on the size of the
+ * pool/queue and the drop probability becomes 100% when the queue/pool
+ * is full.
+ * RED is logically configured in the CoS and could be implemented
+ * in either pool or queue linked to the CoS depending on
+ * platform capabilities. Application should make sure not to link
+ * multiple CoS with different RED or BP configuration to the same queue
+ * or pool.
+ * RED is enabled when the resource limit is equal to or greater than
+ * the maximum threshold value and is disabled when resource limit
+ * is less than or equal to minimum threshold value. */
+typedef struct odp_red_param_t {
+	/** A boolean to enable RED
+	 * When true, RED is enabled and configured with RED parameters.
+	 * Otherwise, RED parameters are ignored. */
+	odp_bool_t red_enable;
+
+	/** Threshold parameters for RED
+	 * RED is enabled when the resource limit is equal to or greater than
+	 * the maximum threshold value and is disabled when resource limit
+	 * is less than or equal to minimum threshold value. */
+	odp_threshold_t red_threshold;
```

Comment:
I think it should be:
    * drop probability == 100%, when resource usage > threshold.max
    * drop probability == 0%, when resource usage < threshold.min

Now the text describe min/max as hysteresis: > max enables, < min disables. Which is not the intention, I guess.

Reply:
The drop probability is 100% only when the pool is completely full i.e there is no further buffer to allocate packet. The intention of this description is that when resource usage is greater than threshold.max then the drop probability is enabled and packets will get dropped on an increasing drop probability and when it is less than min threshold then drop probability is disabled.

Reply:
As we discussed during today's ARCH call. `max` and `min` can be renamed `start` and `stop` to more accurately reflect the actions. RED / backpressure begins when utilization hits the `start` threshold and is deasserted when utilization drops back to the `stop` threshold. So `start` >= `stop`.  If the two are equal it means there is no hysteresis.

The question arises whether a third `max` threshold value should exist which drops / backpressure is held at 100%. The idea here is to reserve a portion of the pool/queue resource for application use independent of use by PktIOs. Since this may not be feasible in all implementations this should probably be advertised with additional capability info.r

Reply:
"Reaches between" is not valid since drop probability is defined also when <min and >max. "... when the packets in the queue/pool reaches between the specified threshold." -> "... when the packets in the queue/pool cross specified threshold values."

The "resource limit" needs to be specified: is it free or used space? Or is it different for pools and queues: free space in a pool, used space in a queue? May be the text is easier to understand with a bullet list:
* drop probability is 100%, when resource usage > threshold.max
* drop probability is 0%, when resource usage < threshold.min
* drop probability is between 0 ... 100%, when resource usage is between threshold.min and threshold.max

... and then define what "resource usage" means. Pools: space used, queues packet/bytes in queue

Reply:
red_ prefix can be dropped from the struct field names. When red param is used, the name of the variable is likely red or red_param, etc. So, after removing the prefix it looks like this

red.enable = 1;
red.threshold.packets.min = 1000;

vs 

red.red_enable = 1;
red.red_threshold....
...

Reply:
"... RED is enabled when the resource limit is equal to ..."

This refers still to "resource limit", it should be "resource usage" which is a term that has been define above.